src/main/java/de/tum/in/www1/artemis/service/ProgrammingSubmissionService.java

```diff
@@ -138,9 +139,32 @@ public ProgrammingSubmission getLatestPendingSubmission(Long participationId) th
         Optional<ProgrammingSubmission> submissionOpt = programmingSubmissionRepository.findFirstByParticipationIdOrderBySubmissionDateDesc(participationId);
         if (!submissionOpt.isPresent() || submissionOpt.get().getResult() != null) {
             // This is not an error case, it is very likely that there is no pending submission for a participation.
-            return null;
+            return Optional.empty();
         }
-        return submissionOpt.get();
+        return submissionOpt;
+    }
+
+    /**
+     * For every student participation of a programming exercise, try to find a pending submission.
+     *
+     * @param programmingExerciseId for which to search pending submissions
+     * @return a Map of {[participationId]: ProgrammingSubmission | null}. Will contain an entry for every student participation of the exercise and a submission object if a pending submission exists or null if not.
+     */
+    @Transactional(readOnly = true)
+    public Map<Long, Optional<ProgrammingSubmission>> getLatestPendingSubmissionsForProgrammingExercise(Long programmingExerciseId) {
+        Map<Long, Optional<ProgrammingSubmission>> pendingSubmissions = new HashMap<>();
+        List<ProgrammingExerciseStudentParticipation> participations = programmingExerciseParticipationService.findByExerciseId(programmingExerciseId);
+        for (ProgrammingExerciseStudentParticipation participation : participations) {
+            Optional<ProgrammingSubmission> submissionOpt = programmingSubmissionRepository.findFirstByParticipationIdOrderBySubmissionDateDesc(participation.getId());
```

Comment:
I think, you could shorten this, by refactoring the code from `getLatestPendingSubmission` into a private method:
```
private Optional<ProgrammingSubmission> latestPendingSubmissionFor(final long participationId) {
        Optional<ProgrammingSubmission> submissionOpt = programmingSubmissionRepository.findFirstByParticipationIdOrderBySubmissionDateDesc(participationId);
        if (submissionOpt.isEmpty() || submissionOpt.get().getResult() != null) {
            // This is not an error case, it is very likely that there is no pending submission for a participation.
            return Optional.empty();
        }
        
        return submissionOpt;
    }
```
Then you could just write this method in two lines: 
```
public Map<Long, Optional<ProgrammingSubmission>> getLatestPendingSubmissionsForProgrammingExercise(Long programmingExerciseId) {
        List<ProgrammingExerciseStudentParticipation> participations = programmingExerciseParticipationService.findByExerciseId(programmingExerciseId);
        return participations.stream()
            .collect(Collectors.toMap(Participation::getId, p -> latestPendingSubmissionFor(p.getId())));
    }
```

Reply:
Yes, that is probably worth it, I changed the code.